cmd/kubeadm/app/phases/copycerts/copycerts.go

```diff
@@ -159,11 +160,26 @@ func createRBAC(client clientset.Interface) error {
 	})
 }
 
-func getSecretOwnerRef(client clientset.Interface, tokenID string) ([]metav1.OwnerReference, error) {
+func getSecretOwnerRef(client clientset.Interface, tokenID string, isDryRun bool) ([]metav1.OwnerReference, error) {
 	secretName := bootstraputil.BootstrapTokenSecretName(tokenID)
-	secret, err := client.CoreV1().Secrets(metav1.NamespaceSystem).Get(context.TODO(), secretName, metav1.GetOptions{})
-	if err != nil {
-		return nil, errors.Wrap(err, "error to get token reference")
+
+	var err error
+	var secret *v1.Secret
+
+	if isDryRun {
+		// If dry-running, just generate a fake bootstrap token Secret instead of doing a real query
+		secret = &v1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: metav1.NamespaceSystem,
+				Name:      secretName,
+				UID:       uuid.NewUUID(),
+			},
+		}
+	} else {
```

Comment:
i think the better way to do this is modifying the "init getter".
this will result in no need to pass the isDryRun bool to functions in the stack.

here https://github.com/kubernetes/kubernetes/blob/f104ae885fb100278d9cbf2ec3a15b11f465872c/cmd/kubeadm/app/cmd/init.go#L533-L534
kubeadm constructs an init getter and dry run client.

inside the init getter there are some handlers:
https://github.com/kubernetes/kubernetes/blob/b0bc8adbc2178e15872f9ef040355c51c45d04bb/cmd/kubeadm/app/util/apiclient/init_dryrun.go#L61-L64

we can add a new handler for resource type "secret" called something like `handleCopyCertsSecret`
and in the handler we can make sure the secret name matches the prefix as defined here:
https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/cluster-bootstrap/token/util/helpers.go#L109-L111
and return a usable / blank secret object for dryrun.

Reply:
Thank you for your patient explanation @neolit123 

But according to the following code:
```
// handleGetBootstrapToken handles the case where kubeadm init creates the default token; and the token code GETs the
// bootstrap token secret first in order to check if it already exists
func (idr *InitDryRunGetter) handleGetBootstrapToken(action core.GetAction) (bool, runtime.Object, error) {
	if !strings.HasPrefix(action.GetName(), "bootstrap-token-") || action.GetNamespace() != metav1.NamespaceSystem || action.GetResource().Resource != "secrets" {
		// We can't handle this event
		return false, nil, nil
	}
	// We can safely return a NotFound error here as the code will just proceed normally and create the Bootstrap Token
	return true, nil, apierrors.NewNotFound(action.GetResource().GroupResource(), "secret not found")
}
```
It looks like the handler will always return a `NotFound` error whether or not the bootstrap token Secret has already been created. It's hard to decide when to return the `NotFound` error and when to return the fake Secret.

And we never actually store objects to the backend, because of :
https://github.com/kubernetes/kubernetes/blob/0d46aee44eeb4062584469144ff6d8a543a5d1fa/cmd/kubeadm/app/util/apiclient/dryrunclient.go#L162-L166
https://github.com/kubernetes/kubernetes/blob/0d46aee44eeb4062584469144ff6d8a543a5d1fa/cmd/kubeadm/app/util/apiclient/dryrunclient.go#L199-L206

Another approach might be to allow the bootstrap token Secret to be stored on the backend and then remove the handleGetBootstrapToken handler.

Reply:
I tried the above approach. It also works. @neolit123